src/main/scala/akka/persistence/jdbc/query/scaladsl/JdbcReadJournal.scala

```diff
@@ -135,39 +134,63 @@ class JdbcReadJournal(config: Config)(implicit val system: ExtendedActorSystem)
+        }
+    }
+  }
+
+  /**
+    * @param terminateAfterOffset If None, the stream never completes. If a Some, then the stream will complete once a
+    *                             query has been executed which might return an event with this offset (or a higher offset).
+    *                             The stream may include offsets higher than the value in terminateAfterOffset, since the last batch
+    *                             will be returned completely.
+    */
+  private def eventsByTag(tag: String, offset: Long, terminateAfterOffset: Option[Long]): Source[EventEnvelope, NotUsed] = {
     import akka.pattern.ask
-    implicit val askTimeout = Timeout(readJournalConfig.journalSequenceRetrievalConfiguration.askTimeout)
-    Source.fromFuture(orderingActor.ask(GetMaxOrderingId).mapTo[MaxOrderingId])
-      .flatMapConcat { latestOrdering =>
-        if (latestOrdering.maxOrdering < offset) Source.empty
-        else readJournalDao.eventsByTag(tag, offset, latestOrdering.maxOrdering, max)
-      }
-      .mapAsync(1)(Future.fromTry)
+    implicit val askTimeout: Timeout = Timeout(readJournalConfig.journalSequenceRetrievalConfiguration.askTimeout)
+    /* We unfold with 3 parameters:
+     * - The first parameter (from) determines the starting offset (i.e. the minimum value) to query from.
+     * - The second parameter (completeImmediately) signifies that we have queried events until at least
+     *   `terminateAfterOffset`, therefore the stream no longer needs to continue.
+     * - The third parameter (retrieveImmediately) is a boolean which determines whether the database query should be
+     *   executed immediately, or after a delay.
+     */
+    Source.unfoldAsync[(Long, Boolean, Boolean), Seq[EventEnvelope]]((offset, false, false)) {
+      case (from: Long, completeImmediately: Boolean, retrieveImmediately: Boolean) =>
+        def retrieveNextBatch(): Future[Some[((Long, Boolean, Boolean), Seq[EventEnvelope])]] = {
+          for {
+            queryUntil <- journalSequenceActor.ask(GetMaxOrderingId).mapTo[MaxOrderingId]
+            xs <- currentJournalEventsByTag(tag, from, readJournalConfig.maxBufferSize, queryUntil).runWith(Sink.seq)
+          } yield {
+            val isFullBatch = xs.size == readJournalConfig.maxBufferSize
+            val completeStreamAfterThisBatch = !isFullBatch && terminateAfterOffset.exists(_ <= queryUntil.maxOrdering)
+            val nextStartingOffset = if (xs.isEmpty) from else xs.map(_.offset.value).max
+            Some((nextStartingOffset, completeStreamAfterThisBatch, isFullBatch), xs)
+          }
+        }
+        if (completeImmediately) Future.successful(None)
+        else {
+          if (retrieveImmediately) retrieveNextBatch()
+          else akka.pattern.after(readJournalConfig.refreshInterval, system.scheduler)(retrieveNextBatch())
+        }
+    }.mapConcat(identity)
```

Comment:
That's a more general comment. 

I find it hard to reason when using two booleans to express three states. IIUC, we need here a flow control with three states: `Continue`, `ContinueDelayed` and `Stop`. 
 
I suggest the following:
```scala

sealed trait FlowControl
    
/** Keep querying - used when we are sure that there is more events to fetch */
object Continue extends FlowControl

/** Keep querying with delay - used when we have consumed all events,
 * but want to poll for future events */
object ContinueDelayed extends FlowControl

/** Stop querying - used when we reach the desired offset  */
object Stop extends FlowControl


Source.unfoldAsync[(Long, FlowControl), Seq[EventEnvelope]]((offset, Continue)) {

  case (from, control) =>

    val batchSize = readJournalConfig.maxBufferSize

    def retrieveNextBatch() = {

      for {
        queryUntil <- journalSequenceActor.ask(GetMaxOrderingId).mapTo[MaxOrderingId]
        xs <- currentJournalEventsByTag(tag, from, batchSize, queryUntil).runWith(Sink.seq)
      } yield {

        def decideHowToContinue = {
          val rangeSize = queryUntil - from
          val expectedBatchSize = if (rangeSize < batchSize) rangeSize else batchSize

          // if we fetch less than expected it means we reached the end of the stream
          // we want to keep querying, but adding some delay
          if (xs.size < expectedBatchSize) ContinueDelayed
          else Continue
        }

        val control =
          terminateAfterOffset
            // keep it as Some if target is reached
            .filterNot(_ > queryUntil.maxOrdering)
            // we can stop because we have reached the target offset
            .map(_ => Stop)
            // othewise decide how to continue
            .getOrElse(decideHowToContinue)

        val nextStartingOffset = if (xs.isEmpty) from else xs.map(_.offset.value).max

        Some((nextStartingOffset, control), xs)
      }
    }

    control match {
      case Stop => Future.successful(None)
      case Continue => retrieveNextBatch()
      case ContinueDelayed =>
        akka.pattern.after(readJournalConfig.refreshInterval, system.scheduler)(retrieveNextBatch())
    }

}.mapConcat(identity)
```

NOTE: this implementation resolves the points I mentioned before:
 * start without delay
 * isFullBatch fix
 * stop as soon as we detected we reach the target

Reply:
Looks good. However I think we should introduce a in case the size of the batch is range size (as I mentioned in my other comment)

Reply:
Maybe to make clear that we have indeed fetched the targeted offset...
```scala

val offsetsOnly = xs.map(_.offset.value)
val control =
  terminateAfterOffset
    // keep it as Some if target is reached
    .filter(target => offsetsOnly.contains(target))
    // we can stop because we have reached the target offset
    .map(_ => Stop)
    // otherwise we must decide how to continue
    .getOrElse(decideHowToContinue)
```